blindsidenetworks/scalelite
# SL-244: Add voice bridge allocation & freeswitch dialplan generation. (closes #769)

## Description
To support the ability to have POTS callers dial in to meetings handled by the scalelite load balancer, there has to be a system in place to look up the voice bridge number, and direct the call to the correct backend bigbluebutton server.

This is done by having scalelite handle allocating unique voice bridge numbers for each meeting internally, and having an index to allow looking up a meeting from its voice bridge number. Configuration options are provided to control the length of voice bridge numbers to allocate, and to choose whether or not to use voice bridge numbers provided by external integrations (as long as they're unique).

The actual dial in system itself is handled by having a freeswitch server (separate from the BBB servers) that receives external calls, and makes dialplan api calls (using mod_xml_curl) to an endpoint on scalelite. This endpoint dynamically generates dialplans that prompt for the voice bridge number and forward the call to the correct BigBlueButton server.

## Testing Steps
For the voice bridge allocation, no special test setup is required. The new configuration options are documented in the README.md, and you can create and join meetings and verify that the conference number displayed (e.g. in the welcome message) has the desired properties.

Setup for a freeswitch server for testing the dialplan and forwarding documentation is to be written.

## Notes

A lot of the "create"-related tests previously used Webmock with hard-coded URLs which no longer work when the dynamically generated voiceBridge parameter is added. These tests have all been rewritten to test for the specific parameters required on the call rather than a hardcoded URL.

Rather than work out how to do the same changes on the rspec tests, I have reimplemented the (very small number of) rspec tests in minitest and removed rspec. It's better to only have one test framework instead of 2, anyways.
